[LBSE] Fix RenderSVGResourceClipper introduction regressions

rwlbuis opened the pull request:
#### 5d9e979451aa984f77f416ae8f3a98b86373f081
<pre>
[LBSE] Fix RenderSVGResourceClipper introduction regressions
<a href="https://bugs.webkit.org/show_bug.cgi?id=263650">https://bugs.webkit.org/show_bug.cgi?id=263650</a>

Reviewed by Nikolas Zimmermann.

This patch fixes regressions introduced by r269635, I falsely relied on buildbot, but
LBSE needs some manual checking.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintSVGResourceLayer): remove incorrect save() call.
(WebCore::RenderLayer::paintLayerContents): test correct paint flags variable for PaintingSVGClippingMask.
(WebCore::RenderLayer::hitTestLayer): add logic to return false for  hit test of resources and hidden containers.
* Source/WebCore/rendering/RenderLayerInlines.h:
(WebCore::RenderLayer::hasNonOpacityTransparency const): logic was inverted.
* Source/WebCore/rendering/svg/SVGResources.cpp:
(WebCore::SVGResources::buildCachedResources): do not add pending resource for LBSE clippers.

Canonical link: <a href="https://commits.webkit.org/269774@main">https://commits.webkit.org/269774@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b7d281ee0e4099b7c6dd2633d5f57871d0f244a8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23557 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1674 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24679 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25718 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21750 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23826 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3237 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24097 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23799 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1222 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/20383 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26314 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/1020 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21282 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/21477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21547 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25319 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/18710 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/968 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5615 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/1409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1276 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/rendering/svg/SVGResources.cpp:
```diff
@@ -232,7 +232,7 @@ std::unique_ptr<SVGResources> SVGResources::buildCachedResources(const RenderEle
             if (auto* clipper = getRenderSVGResourceById<LegacyRenderSVGResourceClipper>(treeScope, id))
                 ensureResources(foundResources).setClipper(clipper);
-            else
+            else if (!renderer.document().settings().layerBasedSVGEngineEnabled())
```
I guess this is temporary, since we don't want LBSE to use this SVGResources system.

rwlbuis replied:
Yes, a quick fix for testing, we will need some/similar logic here for every LBSE resource we introduce.

nikolaszimmermann replied:
Great!

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayerInlines.h:
```diff
@@ -61,7 +61,7 @@ inline bool RenderLayer::hasNonOpacityTransparency() const
 
 #if ENABLE(LAYER_BASED_SVG_ENGINE)
-    if (renderer().document().settings().layerBasedSVGEngineEnabled())
+    if (!renderer().document().settings().layerBasedSVGEngineEnabled())
```
:-)

rwlbuis replied:
think you pointed this one out yesterday.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -2929,11 +2929,10 @@ static inline bool shouldSuppressPaintingLayer(RenderLayer* layer)
 }
 
-void RenderLayer::paintSVGResourceLayer(GraphicsContext& context, GraphicsContextStateSaver& stateSaver, const AffineTransform& layerContentTransform)
+void RenderLayer::paintSVGResourceLayer(GraphicsContext& context, GraphicsContextStateSaver&, const AffineTransform& layerContentTransform)
```
Time to nuke the GraphicsContextStateSaver parameter.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3374,6 +3373,7 @@ void RenderLayer::paintLayerContents(GraphicsContext& context, const LayerPainti
     GraphicsContextStateSaver stateSaver(context, false);
     RegionContextStateSaver regionContextStateSaver(paintingInfo.regionContext);
 
+    localPaintFlags.remove(PaintLayerFlag::PaintingSVGClippingMask);
```
I'd suggest to directly remove the PaintingSVGClippingMask from 'localPaintFlags' after we called setupClipPath() - this guarantees that we have no side effects and never propagate the PaintiningSVGClippingMask flags down the tree.

nikolaszimmermann replied:
Something like this:

```
diff --git a/Source/WebCore/rendering/RenderLayer.cpp b/Source/WebCore/rendering/RenderLayer.cpp
index 96f6d76acafb..4244e181b157 100644
--- a/Source/WebCore/rendering/RenderLayer.cpp
+++ b/Source/WebCore/rendering/RenderLayer.cpp
@@ -3373,10 +3373,13 @@ void RenderLayer::paintLayerContents(GraphicsContext& context, const LayerPainti
     GraphicsContextStateSaver stateSaver(context, false);
     RegionContextStateSaver regionContextStateSaver(paintingInfo.regionContext);
 
-    localPaintFlags.remove(PaintLayerFlag::PaintingSVGClippingMask);
     if (shouldApplyClipPath(paintingInfo.paintBehavior, localPaintFlags))
         setupClipPath(context, stateSaver, regionContextStateSaver, paintingInfo, localPaintFlags, columnAwareOffsetFromRoot);
 
+    bool applySVGClippingMask = localPaintFlags.contains(PaintLayerFlag::PaintingSVGClippingMask);
+    if (applySVGClippingMask)
+        localPaintFlags.remove(PaintLayerFlag::PaintingSVGClippingMask);
+
     bool selectionAndBackgroundsOnly = paintingInfo.paintBehavior.contains(PaintBehavior::SelectionAndBackgroundsOnly);
     bool selectionOnly = paintingInfo.paintBehavior.contains(PaintBehavior::SelectionOnly);
 
@@ -3512,7 +3515,7 @@ void RenderLayer::paintLayerContents(GraphicsContext& context, const LayerPainti
             paintMaskForFragments(layerFragments, context, paintingInfo, paintBehavior, subtreePaintRootForRenderer);
         }
 
-        if (localPaintFlags & PaintLayerFlag::PaintingSVGClippingMask || (!(paintFlags & PaintLayerFlag::PaintingCompositingMaskPhase) && (paintFlags & PaintLayerFlag::PaintingCompositingClipPathPhase))) {
+        if (applySVGClippingMask || (!(paintFlags & PaintLayerFlag::PaintingCompositingMaskPhase) && (paintFlags & PaintLayerFlag::PaintingCompositingClipPathPhase))) {
             // Re-use paintChildClippingMaskForFragments to paint black for the compositing clipping mask.
             paintChildClippingMaskForFragments(layerFragments, context, paintingInfo, paintBehavior, subtreePaintRootForRenderer);
         }
```

nikolaszimmermann reviewed (approved):
r=me, with a minor change.

rwlbuis committed:
[LBSE] Fix RenderSVGResourceClipper introduction regressions
https://bugs.webkit.org/show_bug.cgi?id=263650

Reviewed by Nikolas Zimmermann.

This patch fixes regressions introduced by r269635, I falsely relied on buildbot, but
LBSE needs some manual checking.

* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintSVGResourceLayer): remove incorrect save() call.
(WebCore::RenderLayer::paintLayerContents): test correct paint flags variable for PaintingSVGClippingMask.
(WebCore::RenderLayer::hitTestLayer): add logic to return false for  hit test of resources and hidden containers.
* Source/WebCore/rendering/RenderLayerInlines.h:
(WebCore::RenderLayer::hasNonOpacityTransparency const): logic was inverted.
* Source/WebCore/rendering/svg/SVGResources.cpp:
(WebCore::SVGResources::buildCachedResources): do not add pending resource for LBSE clippers.

Canonical link: https://commits.webkit.org/269774@main
